plugins/inputs/ctrlx_datalayer_sse/ctrlx_datalayer_sse.go

```
@@ -0,0 +1,498 @@
+
+	pl := sub.payload()
+	p, _ := json.Marshal(pl)
+
+	requestBody := bytes.NewBuffer(p)
+	req, err := http.NewRequest("POST", sseURL, requestBody)
+	if err != nil {
+		s.Log.Errorf("Failed to create post request to create sse subscription %v, error: %v", sub.index, err)
+		return ""
+	}
+
+	req.Header.Add("Authorization", s.token.String())
+
+	resp, err := s.connection.Do(req)
+	if err != nil {
+		s.Log.Errorf("Failed to do post request to create sse subscription %v, error: %v", sub.index, err)
+		return ""
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 && resp.StatusCode != 201 {
+		s.Log.Errorf("Failed to create sse subscription %v, status: %v", sub.index, resp.Status)
+		return ""
+	}
+
+	return sseURL + "/" + pl.Properties.ID
+}
+
+// reconnect is called when the connection is broken
+// reconnect starts restablishing the connection
+func (s *ctrlXDataLayerSSE) reconnect(sub subscription) {
+	go func() {
+		s.Log.Infof("Subscription %v is down, reconnect is starting", sub.index)
+		for {
+			time.Sleep(time.Duration(defaultReconnectInterval))
+
+			if !s.setAuthToken() {
+				continue
+			}
+
+			subURL := s.createSubscription(&sub)
+			if subURL == "" {
+				continue
+			}
+
+			client := newClient(subURL, sub, s.Log)
+			if s.autoReconnect {
+				client.onDisconnect(s.reconnect)
+			}
+			client.Headers["Authorization"] = s.token.String()
+			client.Connection.Transport = httpTransport(s.InsecureSkipVerify)
+			err := client.requestData(s, sub.Measurement)
+			if err != nil {
+				continue
+			}
+
+			s.Log.Infof("Subscription %v is reconnected", sub.index)
+			return
+		}
+	}()
+}
+
+// addMetric writes sse metric into accumulator
+func (s *ctrlXDataLayerSSE) addMetric(se *sSEEvent, measurement string) {
+	m := metric{}
+	switch se.Event {
+	case "update":
+		// Received an updated value, that we translate into a metric
+		_ = json.Unmarshal([]byte(se.Data), &m)
+		t := convertTimestamp2UnixTime(m.Timestamp)
+		if s.acc != nil {
```

How can `acc` be `nil` here? Even if it is possible, this will silently drop the metric!

---

The function `addMetric` might be called after the service stop function is called, as it is running in separate thread. `acc` is set to `nil` in stop function. Telegraf --config example.conf --test this command can simulate this issue very often.